.github/copilot-instructions.md

```diff
@@ -0,0 +1,204 @@
+
+## Key Architecture Concepts
+
+### System Hierarchy
+```
+IoT System (e.g., ~/iot-systems/my-home/)
+├── system.conf              # System-wide configuration (WiFi, MQTT broker)
+└── node-name/               # Each microcontroller has its own folder
+    ├── node.conf            # Board type and node-specific settings
+    ├── setup.cpp            # Device configuration (sensors/actuators)
+    ├── key.txt              # Security key for OTA updates
+    └── build/               # Auto-generated PlatformIO project (transient)
+```
+
+**Important**: Nodes and devices are defined declaratively in `setup.cpp` using a high-level C++ DSL. The system automatically handles MQTT topics, networking, security, and OTA updates.
+
+## Directory Structure
+
+### Documentation (`/doc`)
+- `architecture.rst` - System architecture overview
+- `first-node.rst` / `second-node.rst` - Essential tutorials
+- `device-architecture.rst` - Device framework details
+- `deployment-process.rst` - Build and deployment workflow
+- `node_help/` - Device command references
+- `projects_help/` - Project configuration help
+
+### System Management (`/bin`)
+- `compile` - Compiles node firmware
+- `deploy` - Deploys firmware (serial or OTA)
+- `command_checker.py` - Detects which device drivers are needed
+- `create_node_template` - Creates new node folders
+- Other gateway management scripts (accesspoint, MQTT, etc.)
+
+### MCU Device Drivers (`/lib/node_types/esp/src`)
+**Critical**: This folder uses an **implicit inheritance structure** to build PlatformIO projects:
+- `device.h/.cpp` - Base Device class (all devices inherit from this)
+- `dev_*.h/.cpp` - Individual device drivers (sensors, actuators, I2C devices, etc.)
+- `devices.ini` - Device metadata and dependencies
+- `main.cpp` - Main loop (handles polling, MQTT, OTA)
+- `setup.cpp` - Template (overridden by user's setup.cpp)
+
+When `deploy` is called, the system:
+1. Analyzes the node's `setup.cpp` to detect which devices are used
+2. Generates `devices_generated.h` with required includes
+3. Assembles a complete PlatformIO project in `node-folder/build/`
+4. Copies necessary source files using the implicit inheritance structure
+5. Compiles and uploads firmware
+
+## Writing setup.cpp Files
+
+### Device Naming Convention
+**Do NOT quote device names** - they come from command auto-expansion:
+```cpp
+// Correct - device names are NOT strings
+output(blue_led, ONBOARDLED).off();
+input(button1, D3, "up", "down");
+analog(temp_sensor).with_threshold(30, "hot", "cold");
+
+// Incorrect - do not use quoted names
+// output("blue_led", ONBOARDLED);  // WRONG
+```
+
+### MQTT is Automatic
+**Do NOT manually configure MQTT** - the system handles it automatically:
+- Each device publishes to `<topic_prefix>/<node_name>/<device_name>`
```

Comment:
topic_prefix is not completely correct - it's the folder hierarchy - for a folder a/b/c/node.conf (c is therefore the device name) a device publishes to a/b/c/device_name/[subtopic] if the device was configured as blue_led it would regularly publish it's status to a/b/c/blue_led and receive switch commands on a/b/c/blue_led/set